Packet++/src/TelnetLayer.cpp

```diff
@@ -0,0 +1,483 @@
+{
+	size_t offset = 0;
+	while (offset < len)
+	{
+		// Move to next field
+		size_t length = getFieldLen(pos, len - offset);
+		pos += length;
+		offset += length;
+
+		if (isCommandField(pos))
+			return pos;
+	}
+
+	return NULL;
+}
+
+int16_t TelnetLayer::getSubCommand(uint8_t *pos, size_t len)
+{
+	if (len < 3 || pos[1] < Subnegotiation)
+		return TelnetOptionNoOption;
+	return pos[2];
+}
+
+uint8_t *TelnetLayer::getCommandData(uint8_t *pos, size_t &len)
+{
+	if (pos[1] == Subnegotiation && len > 3)
+	{
+		len -= 3;
+		return &pos[3];
+	}
+	len = 0;
+	return NULL;
+}
+
+std::string TelnetLayer::getDataAsString(bool removeEscapeCharacters)
+{
+	uint8_t *dataPos = NULL;
+	if (isDataField(m_Data))
+		dataPos = m_Data;
+	else
+		dataPos = getNextDataField(m_Data, m_DataLen);
+
+	if (!dataPos)
+	{
+		PCPP_LOG_DEBUG("Packet does not have a data field");
+		return std::string();
+	}
+
+	// Convert to string
+	if (removeEscapeCharacters)
+	{
+		std::stringstream ss;
+		for (size_t idx = 0; idx < m_DataLen - (dataPos - m_Data) + 1; ++idx)
+		{
+			if (int(dataPos[idx]) < 127 && int(dataPos[idx]) > 31) // From SPACE to ~
+				ss << dataPos[idx];
+		}
+		return ss.str();
+	}
+	return std::string((char *)m_Data, m_DataLen);
+}
+
+size_t TelnetLayer::getTotalNumberOfCommands()
+{
+	size_t ctr = 0;
+	if (isCommandField(m_Data))
+		++ctr;
+
+	size_t offset = 0;
+	uint8_t *pos = m_Data;
+	while (pos != NULL)
+	{
+		offset = pos - m_Data;
+		pos = getNextCommandField(pos, m_DataLen - offset);
+		if (pos)
+			++ctr;
+	}
+
+	return ctr;
+}
+
+size_t TelnetLayer::getNumberOfCommands(TelnetCommand command)
+{
+	if (command < 0)
+		return 0;
+
+	size_t ctr = 0;
+	if (isCommandField(m_Data) && m_Data[1] == command)
+		++ctr;
+
+	size_t offset = 0;
+	uint8_t *pos = m_Data;
+	while (pos != NULL)
+	{
+		offset = pos - m_Data;
+		pos = getNextCommandField(pos, m_DataLen - offset);
+		if (pos && pos[1] == command)
+			++ctr;
+	}
+
+	return ctr;
+}
+
+TelnetLayer::TelnetCommand TelnetLayer::getFirstCommand()
+{
+	// If starts with command
+	if (isCommandField(m_Data))
+		return static_cast<TelnetCommand>(m_Data[1]);
+
+	// Check is there any command
+	uint8_t *pos = getNextCommandField(m_Data, m_DataLen);
+	if (pos)
+		return static_cast<TelnetCommand>(pos[1]);
+	return TelnetCommandEndOfPacket;
+}
+
+TelnetLayer::TelnetCommand TelnetLayer::getNextCommand()
+{
+	if (lastPositionOffset == UINT16_MAX)
+	{
+		lastPositionOffset = 0;
+		if (isCommandField(m_Data))
+			return static_cast<TelnetLayer::TelnetCommand>(m_Data[1]);
+	}
+
+	uint8_t *pos = getNextCommandField(&m_Data[lastPositionOffset], m_DataLen - lastPositionOffset);
+	if (pos)
+	{
+		lastPositionOffset = pos - m_Data;
+		return static_cast<TelnetLayer::TelnetCommand>(pos[1]);
+	}
+	lastPositionOffset = UINT16_MAX;
+	return TelnetCommandEndOfPacket;
+}
+
+TelnetLayer::TelnetOption TelnetLayer::getOption()
+{
+	if (lastPositionOffset <= m_DataLen)
```

Comment:
this should be `<` instead of `<=`, no? 🤔

Reply:
Definitely